ncmpcpp: add module

olmokramer opened the pull request:
### Description

<!--

Please provide a brief description of your change.

-->

Adds a module for https://github.com/ncmpcpp/ncmpcpp. All settings are configured through a single `settings` attrset, except for `mpd_music_dir`, which has its own dedicated option. If it is set to `null` (the default) its value is taken from `services.mpd.musicDirectory` and `services.mpd.enable` is required to be `true`.

### Checklist

<!--

Please go through the following checklist before opening a non-WIP
pull-request.

Also make sure to read the guidelines found at

  https://github.com/rycee/home-manager/blob/master/doc/contributing.adoc#sec-guidelines

-->

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all`.

- [x] Test cases updated/added. See [example](https://github.com/rycee/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/rycee/home-manager/blob/master/doc/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/rycee/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [x] Added myself as module maintainer. See [example](https://github.com/rycee/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [x] Added myself and the module files to `.github/CODEOWNERS`.

olmokramer commented on modules/programs/ncmpcpp.nix:
```diff
@@ -0,0 +1,142 @@
+      example = "pkgs.ncmpcpp.override { ... }";
+    };
+
+    mpdMusicDir = mkOption {
```
We could get rid of this option and place it under `settings` if `settings` were a [freeform module](https://nixos.org/manual/nixos/unstable/#sec-freeform-modules), but that breaks the tests with an error:

```
Module `:anon-1:anon-1' has an unsupported attribute `freeformType'.
This is caused by introducing a top-level `config' or `options'
attribute. Add configuration attributes immediately on the top level
instead, or move all of them (namely: freeformType) into the
explicit `config' attribute.
```

The [Freeform modules PR](https://github.com/NixOS/nixpkgs/pull/82743) has only been merged 12 days ago, though, so it may not be a great idea to start relying on it already.

olmokramer replied:
This is what that might look like https://github.com/olmokramer/home-manager/blob/ncmpcpp-freeform/modules/programs/ncmpcpp.nix

rycee replied:
Cool. I'll have to check out 🙂

rycee commented on modules/programs/ncmpcpp.nix:
```diff
@@ -0,0 +1,142 @@
+    };
+
+    bindings = mkOption {
+      type = types.listOf bindingType;
```
Perhaps could be `with types; attrsOf (either str (listOf str))`? I.e., the example would be
```nix
{
  "j" = "scroll_down";
  "k" = "scroll_up";
  "J" = [ "select_item" "scroll_down" ];
  "K" = [ "select_item" "scroll_up" ];
}
```

olmokramer replied:
That's what I tried initially, but you can bind the same key multiple times in `~/.config/ncmcpcpp/bindings`. For example, I have this in my config:

```
def_key "x"
  delete_playlist_items
def_key "x"
  delete_browser_items
def_key "x"
  delete_stored_playlist
```

rycee replied:
Fair enough.

rycee commented on modules/programs/ncmpcpp.nix:
```diff
@@ -0,0 +1,142 @@
+          + " use services.mpd.musicDirectory.";
+      }
+      {
+        assertion = !(cfg.settings ? mpd_music_dir);
```
I'm not sure it's nice to force the user to specify a music directory. I've been using ncmpcpp for many years without setting this. I always use it with a mpd running on a different host.

olmokramer replied:
Ah I hadn't thought about that case. In that case it's definitely not nice to force this. I'll remove the check :)

rycee commented on modules/programs/ncmpcpp.nix:
```diff
@@ -0,0 +1,142 @@
+      '';
+
+      "ncmpcpp/bindings" = lib.mkIf (lib.length cfg.bindings > 0) {
+        text = ''
```
IMHO it would look slightly nicer with `text = renderBindings cfg.bindings + "\n";`

olmokramer replied:
I'll change it

rycee commented on modules/programs/ncmpcpp.nix:
```diff
@@ -0,0 +1,142 @@
+
+    mpdMusicDir = mkOption {
+      type = types.nullOr types.path;
+      default = null;
```
Could use
```nix
if mpdCfg.enable then mpdCfg.musicDirectory else null
```
as the default? If so, then you'll also need to add a `defaultText` field. Something like
```
defaultText = literalExample ''
  if config.services.mpd.enable then
    config.services.mpd.musicDirectory
  else
    null
'';
```
perhaps.

olmokramer replied:
Yeah that's a great idea!

olmokramer commented:
Thanks for the review! I have implemented all your suggestions except for the bindings type, and added a warning if both `programs.ncmpcpp.mpdMusicDir` and `programs.ncmpcpp.settings.mpd_music_dir` are set.

rycee commented on modules/programs/ncmpcpp.nix:
```diff
@@ -0,0 +1,136 @@
+
+  options.programs.ncmpcpp = {
+    enable =
+      mkEnableOption "ncmpcpp - An ncurses Music Player Daemon (MPD) client";
```
```suggestion
      mkEnableOption "ncmpcpp – an ncurses Music Player Daemon (MPD) client";
```

rycee commented on modules/programs/ncmpcpp.nix:
```diff
@@ -0,0 +1,136 @@
+    package = mkOption {
+      type = types.package;
+      default = pkgs.ncmpcpp;
+      defaultText = "pkgs.ncmpcpp";
+      description = "Package providing the <code>ncmpcpp</code> command.";
+      example = "pkgs.ncmpcpp.override { ... }";
```
```suggestion
      defaultText = literalExample "pkgs.ncmpcpp";
      description = ''
        Package providing the <command>ncmpcpp</command> command.
      '';
      example = literalExample "pkgs.ncmpcpp.override { visualizerSupport = true; }";
```

olmokramer commented:
Thanks again! I've applied your changes and added a `with lib;` at the top, since all modules seem to do that and you seem to be expecting that as well :)

rycee commented on modules/programs/ncmpcpp.nix:
```diff
@@ -0,0 +1,136 @@
+
+    mpdMusicDir = mkOption {
+      type = types.nullOr types.path;
+      default = if mpdCfg.enable then mpdCfg.musicDirectory else null;
```
I suspect this also should verify that the target platform is Linux based since it's more of a historic accident that the mpd module is available on Darwin. Ideally we'll remove the module from Darwin builds in the future and when that happens this code would break. Perhaps something like
```suggestion
      default = if pkgs.stdenv.hostPlatform.isLinux && mpdCfg.enable then
        mpdCfg.musicDirectory
      else
        null;
```
would work.

olmokramer replied:
Yeah I also wondered if this would work on MacOS, but I figured that since the MPD service module is available on MacOS that it would simply work. If the service should realistically only be available on Linux this change seems sensible to me, since a google search for "ncmpcpp mac osx" gives me some results that implicate that ncmpcpp should be working fine on MacOS, so disabling the entire module on MacOS wouldn't make sense.

rycee commented on tests/modules/programs/ncmpcpp/default.nix:
```diff
@@ -0,0 +1,5 @@
+{
+  ncmpcpp-use-mpd-config = ./ncmpcpp-use-mpd-config.nix;
```
This test should only happen for Darwin targets.

olmokramer replied:
I don't follow. In your [previous comment](https://github.com/rycee/home-manager/pull/1457#discussion_r479776480) you seem to be suggesting that the MPD configuration should _not_ be used if the target is Darwin. But this test checks that the MPD configuration _is_ used, if `programs.ncmpcpp.mpdMusicDir` isn't set directly. So shouldn't this test only happen for non-Darwin targets? Or even Linux-only targets?

rycee replied:
Sorry! Yeah, it should definitely be for Linux. The simplest way I imagine is to move the test to a different directory, like `tests/modules/programs/ncmpcpp-linux`, and add that in the Linux only section of `test/default.nix`.

olmokramer replied:
Right, thanks for the clarification and directions on how to test for Linux only! I'll try to make some time tonight to make the changes.

olmokramer commented:
Ok I've changed the default value of `programs.ncmpcpp.mpdMusicDir` to only use the MPD configuration on Linux platforms. Still not quite sure what you meant with your last comment. 

I've also "resolved" the comments that I have addressed for my own convenience, but feel free to "unresolve" them if you think they deserve more attention :)

olmokramer commented:
I've moved the test of the MPD configuration to a new directory and only run it on Linux now.

Also rebuilt the documentation one last time, and I noticed that the default for `programs.ncmpcpp.mpdMusicDir` is not rendered literally, i.e. the newlines and indentation are removed. Is there some alternative to `literalExample` that is supposed to be used for the default value? Or is the default value text never rendered literally?

rycee commented:
Thanks! Rebased to master in 4b702bf6b79843ebd3d6165362ce17f9a1327c9b.
